File: terminale/architectures/se/ordonnancement-implementation/scripts/ocaml/pile.ml
exception Empty;;
type 'a pile = {mutable tab: 'a list};;
let creer () = {tab = []};;
let empiler p x = p.tab <- x::p.tab;;
let depiler p = match p.tab with
| [] -> raise Empty
| h::t -> p.tab <- t; h;;

let p = creer();;
empiler p 3;;
empiler p 4;;
print_int (depiler p);;

File: terminale/architectures/se/ordonnancement-implementation/scripts/c/pile.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Pile Pile;
struct Pile {
  int val;
  Pile *prec;
};

void empiler(Pile **p, int val) {
  Pile *nouv = malloc(sizeof(Pile));
  if (nouv == NULL) {
    EXIT_FAILURE;
  }
  nouv->val = val;
  nouv->prec = *p;
  *p = nouv;
}

int depiler(Pile **p) {
  int x = -1;
  if (*p != NULL) {
    Pile *temp = (*p)->prec;
    x = (*p)->val;
    free(*p);
    *p = temp;
  }
  return x;
}

int main(int argc, char *argv[]) {
  Pile *p = NULL;
  empiler(&p, 5);
  printf("%d", depiler(&p));
  printf("%d", depiler(&p));

  return 0;
}